lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,472 @@
+
+    ctl.cont()
+    ctl.wait_running()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_tstp_tstp_cont_cont(proc, ctl, attrs):
+    ctl.tstp()
+    ctl.wait_stopped()
+
+    ctl.tstp()
+    ctl.wait_stopped()
+
+    ctl.cont()
+    ctl.wait_running()
+
+    ctl.cont()
+    ctl.wait_running()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_bg_fg(proc, ctl, attrs):
+    ctl.bg()
+    ctl.status()
+    ctl.wait_disabled()
+
+    ctl.fg()
+    ctl.status()
+    ctl.wait_enabled()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_bg_fg_no_termios(proc, ctl, attrs):
+    ctl.bg()
+    ctl.status()
+    ctl.wait_disabled()
+
+    ctl.fg()
+    ctl.status()
+    ctl.wait_disabled_fg()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_fg_bg(proc, ctl, attrs):
+    ctl.fg()
+    ctl.status()
+    ctl.wait_enabled()
+
+    ctl.bg()
+    ctl.status()
+    ctl.wait_disabled()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_fg_bg_no_termios(proc, ctl, attrs):
+    ctl.fg()
+    ctl.status()
+    ctl.wait_disabled_fg()
+
+    ctl.bg()
+    ctl.status()
+    ctl.wait_disabled()
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_v_v(proc, ctl, attrs):
+    ctl.fg()
+    ctl.wait_enabled()
+
+    time.sleep(.1)
+    ctl.write(b'v')
+    time.sleep(.1)
+    ctl.write(b'v')
+    time.sleep(.1)
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_v_v_no_termios(proc, ctl, attrs):
+    ctl.fg()
+    ctl.wait_disabled_fg()
+
+    time.sleep(.1)  # allow some time to NOT print
+    ctl.write(b'v\n')
+    time.sleep(.1)  # allow some time to print
+    ctl.write(b'v\n')
+    time.sleep(.1)
+
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def _mock_shell_integration(proc, ctl, attrs):
+    ctl.fg()             # start in foreground
+    ctl.status()
+    ctl.wait_enabled()
+
+    ctl.bg()             # send to background
+    ctl.tstp()           # immediately stop
+    ctl.status()
+    ctl.wait_stopped()
+    ctl.wait_disabled()
+
+    ctl.cont()           # continue in bg, wait a bit
+    ctl.status()
+    ctl.wait_running()
+
+    ctl.fg()             # bring to foreground
+    ctl.status()
+    ctl.wait_enabled()
+
+    time.sleep(.1)       # allow some time to NOT print
+    ctl.write(b'v')      # enable verbose output
+    ctl.status()
+    time.sleep(.1)       # allow some time to print
+
+    ctl.bg()             # send to background (keep running)
+    ctl.status()
+    ctl.wait_disabled()
+
+    ctl.tstp()           # stop in bg
+    ctl.status()
+    ctl.wait_stopped()
+
+    ctl.cont()           # continue in bg for a bit
+    ctl.status()
+    ctl.wait_running()
+
+    ctl.fg()             # bring to foreground
+    ctl.status()
+    ctl.wait_enabled()
+
+    ctl.write(b'v')      # disable verbose and wait a bit
+    ctl.status()
+
+    ctl.bg()             # back to background
+    ctl.status()
+    ctl.wait_disabled()
+
+
+def _mock_shell_integration_no_termios(proc, ctl, attrs):
```

Comment:
This should mention that it doesn't use `proc` or `attrs` (in order to conform to an API). Or throw a boilerplate comment on this and all associated functions. Or put them in a class and document at the class level.

Reply:
Done